search/lib/api/wfs.js

```diff
@@ -180,71 +163,102 @@ async function findCloudCollections (providerId, collectionConceptIds) {
+    const _params = Object.entries(multiParams).map(([k, v]) => {
+      if (arrayParameters.includes(k)) {
+        return [k, v];
+      } else {
+        return [k, v[0]];
+      }
+    });
+    params = stacExtension.prepare(Object.fromEntries(_params));
   } else if (method === 'POST') {
     params = stacExtension.prepare(request.body);
-    fields = request.body.fields;
   } else {
     throw new Error(`Invalid httpMethod ${method}`);
   }
+  return params;
+}
+
+/**
+ * Fetch a list of granules from CMR.
+ */
+async function getItems (request, response) {
+  const providerId = request.params.providerId;
+  const collectionId = request.params.collectionId;
+  const event = request.apiGateway.event;
+
+  const { fields, ...params } = extractParams(request);
 
   try {
-    // CollectionId Path parameter (e.g., /collections/<collectionId>/items vs /search)
     if (collectionId) {
-      Object.assign(params, { collections: [collectionId] });
+      const cmrCollectionId = cmr.stacIdToCmrCollectionId(providerId, collectionId);
+      if (!cmrCollectionId) {
+        return response
+          .status(404)
+          .json(`Collection [${collectionId}] not found for provider [${providerId}]`);
+      } else {
+        params.collections = [collectionId];
+      }
     }
+
     // convert STAC params to CMR Params
     const cmrParams = await cmr.convertParams(providerId, params);
 
-    let granulesResult;
-    if (settings.cmrStacRelativeRootUrl === '/cloudstac') {
-      // Preserve collection_concept_id and concept_id in cmrParams before deleting.
-      // After checking collection_concept_ids being cloud holding collections, they
-      // will be added back one by one because of POST search request requirement.
-      const collectionConceptIds = cmrParams.collection_concept_id;
-      const conceptIds = cmrParams.concept_id;
-      delete cmrParams.collection_concept_id;
-      delete cmrParams.concept_id;
-
-      // Find all the cloud holding collections applicable
-      // i.e. if collection_concept_ids are present, we will get all the cloud holding collections within these ids.
-      // otherwise, we will get all the cloud holding collections for the provider.
-      const allCloudCollections = await findCloudCollections(providerId, collectionConceptIds);
-      const postSearchParams = new URLSearchParams(cmrParams);
-      if (allCloudCollections.length !== 0) {
-        allCloudCollections.forEach(id => {
-          postSearchParams.append('collection_concept_id', id);
-        });
+    let granulesResult = { granules: [], hits: 0 };
+    const collectionsRequested = _.has(params, 'collections');
+    const validCollections = _.has(cmrParams, 'collection_concept_id');
+    if ((collectionsRequested && validCollections) || (!collectionsRequested)) {
+      // if collections param provided, check that not all were filtered out as invalid
+      if (settings.cmrStacRelativeRootUrl === '/cloudstac') {
```

Comment:
How much work would it be to move the contents of the `if` blocks into discrete functions to handle the different routes? I'm concerned about maintainability of long functions if the logic starts getting wildly different per route.

Reply:
It's enough work I think it deserves it's own ticket. I definitely agree and share concerns about maintainability, there is still quite a bit over overlap between the different modules and it's not clear where some functionality lies. I've been trying to minimize the code in the api/ modules and get them to be lighter weight, but our last few PRs have bloated some of the functions back up a bit.